### PR TITLE
Allow missing data in samples

### PR DESCRIPTION
**Pull request by hyanwong**
For brief review, but not merging yet. Addresses https://github.com/tskit-dev/tsinfer/issues/153

### CODE REVIEW

**Comment by hyanwong** [comment]
Ping @marianne-aspbury - this is not quite ready yet, as the Li & Stevens matching algorithm doesn't deal with missing data properly, but it's a good start, and the test suite passes. Most of the fundamental work is in algorithm.py

**Comment by hyanwong** [comment]
NB: currently failing tests because it relies on tskit v0.2.0. Tests pass fine on my machine with the new tskit, so I'm simply waiting for the next tskit release before seeking review.

File: tsinfer/inference.py [hunk 1]
```diff
@@ -821,25 +827,87 @@ def get_samples_tree_sequence(self):
...
         logger.debug("Adding tree sequence edges")
         tables.edges.clear()
         left, right, parent, child = tsb.dump_edges()
-        if np.all(inference_sites == 0):
-            # We have no inference sites, so no edges have been estimated. To ensure
-            # we have a rooted tree, we add in edges for each sample to an artificial
-            # root.
-            assert left.shape[0] == 0
+        logger.debug("Checking for the left/right extent of all available data")
+        # Store the leftmost and rightmost extent of edges leading to TS nodes
```
Comment (hyanwong) [inline tsinfer/inference.py]: There is some slightly hairy logic starting from this line until line 900, which deals with edges whose children are samples that have missing data on the left and right. I would appreciate talking this through with @jeromekelleher when he returns to Oxford.

**Comment by hyanwong** [comment]
Currently the `locate_mutations_on_tree()` functionality is failing on some inferred tree sequences, because samples that are missing all inference sites get [added to a new "artificial root"](https://github.com/hyanwong/tsinfer/blob/d41d19310baa4902667233017111682a09e86c63/tsinfer/inference.py#L879), which is not attached to anywhere else. If the other samples present at that genomic location already fall in a tree, this creates a location with 2 roots, which isn't properly handled by the parsimony code in [`locate_mutations_on_tree()`](https://github.com/hyanwong/tsinfer/blob/d41d19310baa4902667233017111682a09e86c63/tsinfer/inference.py#L635).

**Comment by hyanwong** [comment]
@jeromekelleher I think this is ready to merge. It contains a few separate fixes, so we might want to split them apart. Quite a few are interdependent, however - e.g. upgrading to tskit 0.2.1 means fixing the uint8/int8 haplotype storage, and fixing the appveyor build, then dealing with the missing data in int8 haplotypes means fixing the map_mutations method, so I haven't tried to disentangle them all.

**Comment by jeromekelleher** [comment]
Can you bring this up to date please @hyanwong? Probably the simplest thing to do is rebase, dropping the commits before 47a0153. It's probably easiest to drop the last commit also, as appveyor has been fixed up on master. You may want to separate out the map_mutations code from missing data handling, since this is quite a bit change in itself worthy of its own PR.

**Comment by hyanwong** [comment]
I forgot that the map_mutations code needs committing before this will pass tests. Could you merge https://github.com/tskit-dev/tsinfer/pull/185 first @jeromekelleher ?

**Comment by jeromekelleher** [comment]
Looks like they're interdependant @hyanwong --- see comments over in #185 for the options.

**Comment by awohns** [comment]
Just to confirm, this will handle missing data in samples, but not missing sites, correct? So if not all the samples have information at a given site this would work, but if the sample_data file does not contain all the sites in an ancestors_tree_sequence the code still would break.

**Comment by hyanwong** [comment]
Yes, comments in the function `restore_tree_sequence_builder` says:

```
        # Make sure that the set of positions in the ancestors tree sequence is
        # identical to the inference sites in the sample data file.
```

So I guess it's not set up for changing the inference sites between the ancestors TS and the `match_samples()`  routine.

**Comment by awohns** [comment]
My work around for that check was to remove sites in the `ancestors_ts` which aren't also present in the `sample_data`. This passes the `np.array_equal(position, sample_data_position)` but will still break on 
` if np.any(pos_map[left] != edges.left):
            raise ValueError("Invalid left coordinates")`

**Comment by hyanwong** [comment]
Yes, it's not so simple to change the inference sites in the middle of the pathway, as the indexes will be all up the spout. I'll open another issue about this.

**Comment by jeromekelleher** [comment]
Let's get a few other PRs merged before revisiting this.

**Comment by jeromekelleher** [comment]
Let's bring this up to date and try to get it merged @hyanwong. Looks like you're using some stuff from the head of tskit. Perhaps we should spin a point release of tskit first to avoid messing with installing from git/duplicating code?

**Comment by hyanwong** [comment]
Yes, I'd love to get this merged, as then my student, who is picking up the HIV project soon, can work with the master branch.

I agree, a point release of tskit would be very useful for me at this juncture. Can I help?

**Comment by jeromekelleher** [comment]
> I agree, a point release of tskit would be very useful for me at this juncture. Can I help?

Great; can you have a look through the [milestone](https://github.com/tskit-dev/tskit/milestone/6) and see what we can knock off quickly, and we can triage other stuff for 0.2.4.

**Comment by jeromekelleher** [comment]
Let's try and get this merged this week @hyanwong. Can we schedule some time to go through this together? Friday would be ideal for me.

**Comment by hyanwong** [comment]
Friday is good for me too. Chat then.

File: docs/inference.rst [hunk 2]
```diff
@@ -199,14 +199,29 @@ The final phase of a ``tsinfer`` inference consists of a number steps:
...
    represent the information at these sites. This is done using the tskit
-   `map_mutations <https://tskit.readthedocs.io/en/latest/python-api.html#tskit.Tree.map_mutations>`_.
-   method.
-   
+   `map_mutations <https://tskit.readthedocs.io/en/latest/python-api.html#tskit.Tree.map_mutations>`_
+   method. Edges in the inferred tree sequence usually extend from one inference site up
+   to (but not including) another, covering the non-inference sites
+   between them. There are two exceptions to this, which allow us to
+   encompass any extra non-inference sites that might lie to the left
+   or right of the set of edges above a particular node:
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply docs/inference.rst]: We should open an issue taking note of what we do with missing flanks.

**Comment by hyanwong** [comment]
The current implementation allows sites with missing data to be used for inference, but doesn't change the matching algorithm, which is why `test_missing_inference_sites` in `tests/test_inference.py` currently fails (NB: this is only an issue when matching samples with missing data). There's no point altering the current sample-matching algorithm to skip missing data sites, as this is going to be obsoleted by https://github.com/tskit-dev/tskit/pull/452 and friends. So after discussion with @jeromekelleher,  this PR is being punted down the line until after the generalised tskit L&S matching algorithm is incorporated into `tsinfer`.

All of the work, apart from the incorporation of missing data into matching, is now in this PR, so it should be relatively easy to merge once the L&S plumbing is done.

**Comment by jeromekelleher** [comment]
@hyanwong, I've rebased this and made minimal changes to make the tests pass. I suggest we merge this and then I'll finish up the process of implementing missing data.

**Comment by hyanwong** [comment]
Perfect, thanks

**Pull request merged**